subsys/fs/littlefs_fs.c

```diff
@@ -0,0 +1,663 @@
+	stat->f_blocks = lfs->cfg->block_count;
+
+	path = fs_impl_strip_prefix(path, mountp);
+
+	fs_lock(fs);
+
+	ssize_t ret = lfs_fs_size(lfs);
+
+	fs_unlock(fs);
+
+	if (ret >= 0) {
+		stat->f_bfree = stat->f_blocks - ret;
+		ret = 0;
+	}
+
+	return lfs_to_errno(ret);
+}
+
+/* Return maximum page size in a flash area.  There's no flash_area
+ * API to implement this, so we have to make one here.
+ */
+struct get_page_ctx {
+	const struct flash_area *area;
+	lfs_size_t max_size;
+};
+
+static bool get_page_cb(const struct flash_pages_info *info, void *ctxp)
+{
+	struct get_page_ctx *ctx = ctxp;
+
+	size_t info_start = info->start_offset;
+	size_t info_end = info_start + info->size - 1U;
+	size_t area_start = ctx->area->fa_off;
+	size_t area_end = area_start + ctx->area->fa_size - 1U;
+
+	/* Ignore pages outside the area */
+	if (info_end < area_start) {
+		return true;
+	}
+	if (info_start > area_end) {
+		return false;
+	}
+
+	if (info->size > ctx->max_size) {
+		ctx->max_size = info->size;
+	}
+
+	return true;
+}
+
+/* Iterate over all page groups in the flash area and return the
+ * largest page size we see.  This works as long as the partition is
+ * aligned so that erasing with this size is supported throughout the
+ * partition.
+ */
+static lfs_size_t get_block_size(const struct flash_area *fa)
+{
+	struct get_page_ctx ctx = {
+		.area = fa,
+		.max_size = 0,
+	};
+	struct device *dev = flash_area_get_device(fa);
+
+	flash_page_foreach(dev, get_page_cb, &ctx);
+
+	return ctx.max_size;
+}
+
+static int littlefs_mount(struct fs_mount_t *mountp)
+{
+	int ret;
+	struct fs_littlefs *fs = mountp->fs_data;
+	unsigned int area_id = (uintptr_t)mountp->storage_dev;
+	struct device *dev;
+
+	LOG_INF("LittleFS version %u.%u, disk version %u.%u",
+		LFS_VERSION_MAJOR, LFS_VERSION_MINOR,
+		LFS_DISK_VERSION_MAJOR, LFS_DISK_VERSION_MINOR);
+
+	if (fs->area) {
+		return -EBUSY;
+	}
+
+	/* Create and take mutex. */
+	k_mutex_init(&fs->mutex);
+	fs_lock(fs);
+
+	/* Open flash area */
+	ret = flash_area_open(area_id, &fs->area);
+	if ((ret < 0) || (fs->area == NULL)) {
+		LOG_ERR("can't open flash area %d", area_id);
+		ret = -ENODEV;
+		goto out;
+	}
+	LOG_DBG("FS area %u at 0x%x for %u bytes",
+		area_id, (u32_t)fs->area->fa_off,
+		(u32_t)fs->area->fa_size);
+
+	dev = flash_area_get_device(fs->area);
+	if (dev == NULL) {
+		LOG_ERR("can't get flash device: %s", fs->area->fa_dev_name);
+		ret = -ENODEV;
+		goto out;
+	}
+
+	BUILD_ASSERT(CONFIG_FS_LITTLEFS_READ_SIZE > 0);
+	BUILD_ASSERT(CONFIG_FS_LITTLEFS_PROG_SIZE > 0);
+	BUILD_ASSERT(CONFIG_FS_LITTLEFS_CACHE_SIZE > 0);
+	BUILD_ASSERT(CONFIG_FS_LITTLEFS_LOOKAHEAD_SIZE > 0);
+	BUILD_ASSERT((CONFIG_FS_LITTLEFS_LOOKAHEAD_SIZE % 8) == 0);
+	BUILD_ASSERT((CONFIG_FS_LITTLEFS_CACHE_SIZE
+		      % CONFIG_FS_LITTLEFS_READ_SIZE) == 0);
+	BUILD_ASSERT((CONFIG_FS_LITTLEFS_CACHE_SIZE
+		      % CONFIG_FS_LITTLEFS_PROG_SIZE) == 0);
+
+	lfs_size_t read_size = CONFIG_FS_LITTLEFS_READ_SIZE;
```

Comment:
Another thing to consider. This Kconfig static configuration disqualifies use cases using littlefs for multiple flash devices with different read/prog properties (i.e. internal MCU flash, and an external one). 

Could be these properties defined somewhere in DTS?

Reply:
That's not really true: these settings affect only in-memory caches and alignment of I/O transactions.  As long as the read and prog sizes are a multiple of the flash read/write alignment requirements (1 for SPI-NOR, 4 generally for internal flash) there's no conflict.  The defaults are selected to reduce RAM requirements.

I did look into providing these through a config structure associated with the storage device, with an eye to supporting bigger sizes for partitions with larger data files.  The complexity of managing/validating/using the configuration was high, and limited experimentation didn't really show significant performance advantages to increasing them beyond the default.  That can be revisted in the future, keeping the Kconfig values as defaults when there is no mount-point--specific override.

These properties really are application relevant and belong in Kconfig.  They can be changed without affecting the ability to use an existing file system, unlike the size of the erase block (if that changes, the filesystem will be reformatted).

The erase block comes from devicetree via the flash layout API and can vary between devices.  I have tested littlefs on both SOC flash and SPI-NOR--based file systems using these values.  You can also have partitions on both flash technologies in the same application.

I don't see a great value in changing the approach at this time.

Reply:
> That's not really true: these settings affect only in-memory caches and alignment of I/O transactions. As long as the read and prog sizes are a multiple of the flash read/write alignment requirements (1 for SPI-NOR, 4 generally for internal flash) there's no conflict. The defaults are selected to reduce RAM requirements.
> 
> I did look into providing these through a config structure associated with the storage device, with an eye to supporting bigger sizes for partitions with larger data files. The complexity of managing/validating/using the configuration was high, and limited experimentation didn't really show significant performance advantages to increasing them beyond the default. That can be revisted in the future, keeping the Kconfig values as defaults when there is no mount-point--specific override.

Even if you were right in most cases, I don't see this kind of inflexibility in configuration API as right way to go.  Moreover, what's the benefit of storing 

```c
struct lfs_config cfg;
```
in RAM ? It has about 72 bytes. For most configurations it should be known at compile time and could be placed in flash. However, there should also be way to create it in runtime of course. The most simple solution would be to introduce config macros and allows to predefine it in Kconfig (for static configurations).

Reply:
So, as I said, I *tried* this and it isn't as easy as you think once you get into the details.

However, I share your concerns that at some point in the future per-filesystem configurability might be worthwhile, and the way it will have to be done would require a change to the way applications define littlefs partitions.

OK.  I'll take another stab at it.

Reply:
This is done.  It's not too painful to use the defaults, though it's necessarily different from all the other file systems.

There seems to be a bug in littlefs, related to ARMmbed/littlefs#224, where using a `cache_size` of 512 or larger causes an infinite loop when trying to write a block of data larger than the cache size to an empty file: it writes one cache block of data to each filesystem block until it runs out of disk space.  Apparently the code that transfers from inline files (ones with less than 512 bytes of content) to outline files mishandles this.

As such, the maximum I/O transaction size that works is 256, and the performance improvement from non-default configuration is not great, at least not with a SPI NOR back-end.